Feature #3036: Window Spooling

hawkfish opened the pull request:
*  Add a Seek method for built on a new PrevScanIndex method to support random access to collections. The scans are linear for now, but that should be good enough for the current task.
* Plumb through everything needed to accumulate WindowDataChunk tuples in a ColumnDataCollection.
* Convert the value storage for the "value" window functions to use collections. Note that the validity masks for IGNORE NULLS and EXCLUDE are still memory resident, but they are a lot smaller.
* Create an explicit class for WindowAggregatorLocalState that can handle building and reading the argument collection in parallel instead of using a DataChunk and locks.
* Convert the naïve aggregator to use collections. 
* Convert the segment tree aggregator to use collections.
* Move all of the WindowDataChunk collection scanning functionality into a single class.
* Convert WindowDistinctAggregator to use paging collections.
* Create a wrapper for multi-threaded appending to WindowDataChunk objects.
* Move RANGE values to a collection.
* Convert the custom window functions to use collections instead of in-memory DataChunks.
* Track the insert data validity and pass it down to avoid checks.
* Pass down context so we can get real memory flushing sizes.

Commit:
Feature #3036: ColumnDataCollection Seek

Add a Seek method for built on a new PrevScanIndex method
to support random access to collections.
The scans are linear for now, but that should be good enough
for the current task.

Commit:
Feature #3036: WindowDataChunk With ColumnDataCollection

Plumb through everything needed to accumulate WindowDataChunk tuples
in a ColumnDataCollection. Not used yet.

Commit:
Feature #3036: WindowValueExecutor Spooling

Convert the value storage for the "value" window functions
to use collections. Note that the validity masks for
IGNORE NULLS and EXCLUDE are still memory resident,
but they are a lot smaller.

Commit:
Feature #3036: WindowAggregatorLocalState

Create an explicit class for this that can handle building and reading
the argument collection in parallel instead of using a DataChunk and locks.

Commit:
Feature #3036: WindowNaiveAggregator

Convert the naïve aggregator to use collections.
Also fix loop invariant mistake in PrevScanIndex.

Commit:
Feature #3036: WindowSegmentTree Collections

Convert the segment tree aggregator to use collections.

Commit:
Merge branch 'feature' into window-spooling

Commit:
Feature #3036: WindowTable Consolidation

Move all of the WindowDataChunk collection scanning functionality
into a single class.

Commit:
Feature #3036: WindowDistinctAggregator Collections

Convert WindowDistinctAggregator to use paging collections.

Commit:
Feature #3036: WindowBuilder Appending Wrapper

Create a wrapper for multi-threaded appending to WindowDataChunk objects.

Commit:
Merge branch 'feature' into window-spooling

Commit:
Feature #3036: Window Range Spooling

Move RANGE values to a collection.

Commit:
Merge branch 'feature' into window-spooling

Commit:
Feature #3036: Custom Unary Window

Remove this helper as it relies on DataChunks.

Commit:
Feature #3036: Custom Window Collections

Convert the custom window functions to use collections instead of in-memory DataChunks.

Commit:
Merge branch 'feature' into window-spooling

Commit:
Feature #3036: Custom Window Validity

Track the insert data validity and pass it down to avoid checks.
Also pass down context so we can get real memory flushing sizes.

Commit:
Merge branch 'feature' into window-spooling

lnkuiper commented on src/common/types/column/column_data_collection.cpp:
```diff
@@ -923,6 +923,28 @@ bool ColumnDataCollection::NextScanIndex(ColumnDataScanState &state, idx_t &chun
+	segment_index = state.segment_index;
+	chunk_index = state.chunk_index - 1;
+	state.next_row_index = state.current_row_index;
+	row_index = state.current_row_index -= segments[state.segment_index]->chunk_data[chunk_index].count;
```
Can we split this assignment up into two lines for readability?

lnkuiper reviewed (changes requested):
Code changes look good, this is awesome! So `mad`, `mode` and `quantile` can now be computed on partitions if that single partition is larger than memory? Could you add a test for this? What if we have an evil query with multiple like this:
```sql
SELECT mad(c0) OVER (<window with one large partition that doesn't fit in memory>),
       mode(c0) OVER (<same window>),
       quantile(c0, 0.5) OVER (<same window>)
FROM tbl;
```
Could you add a test for that?

hawkfish committed:
Merge branch 'feature' into window-spooling

hawkfish committed:
Feature #3036: Test Custom Spooling

Add tests to validate that the custom aggregators work with spooling.
Also fix PR feedback and add missing PrepareMergeStage call.

lnkuiper reviewed (approved):
Thanks for the changes! The test looks great :)

Mytherin commented:
Thanks!
